Update SIP-311 tech spec (#1581)
Synthetixio/SIPs: content/sips/sip-311.md

<!--This is where you explain the reasoning behind how you propose to solve the problem. Why did you propose to implement the change in this way, what were the considerations and trade-offs. The rationale fleshes out what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work. The rationale may also provide evidence of consensus within the community, and should discuss important objections or concerns raised during discussion.-->

We propose making the transfer function available on the snxUSD proxy, rather than the core system proxy. This is an extension of the ERC-20 interface (which already essentially includes a "same-chain" `transfer` function).

#### Using CCIP for token transfers

In addition to the new functions outlined in the technical specification below, this SIP also proposes that Chainlink's CCIP token pools be granted permission to mint and burn snxUSD in order to facilitate the cross-chain token transfers. This method allows for multiple benefits as described below:

**Composability & Commercial Opportunity**

> Token transfers using the CCIP token pools provide Synthetix the opportunity, if desired, to open up Chainlink's private snxUSD Token Pool which would allow third parties to transfer snxUSD (even in combination with other tokens and/or data).

Reviewer: ```suggestion
Token transfers using the CCIP token pools provide Synthetix the opportunity in the future, if desired, to open up Chainlink's private snxUSD Token Pool which would allow third parties to transfer snxUSD (even in combination with other tokens and/or data). In this scenario, there is also potential to set up a rev-share model between CCIP and Synthetix (as snxUSD token issuer) for third-party CCIP snxUSD token transfers, creating a shared incentive for increased snxUSD token transfer volume over CCIP.
```

**Enhanced security via the Active Risk Management (ARM) Network**

> With token transfers, the ARM Network has access to additional transaction information (i.e. token, amount, destination chain id) that enable it to perform additional risk management policies which cannot be implemented for arbitrary messaging. The ARM rule set will evolve over time into more advanced logic to identify and mitigate risk more effectively.

Reviewer: ```suggestion
With token transfers, the ARM Network has access to additional transaction information (i.e. token, amount) that enable it to perform additional risk management policies which cannot be implemented for arbitrary messaging. The ARM rule set will evolve over time into more advanced logic to identify and mitigate risk more effectively.
```

**Enhanced security via Rate Limits**

In contrast to arbitrary messaging, CCIP token transfers provide configurable rate limits per token pool per lane, which are set up in alignment with the token issuer. This feature is part of the heavily audited CCIP code base and only available for token transfers. CCIP token transfers also benefit from increased security via the global rate limits (across token pools), so not every token’s rate limit could be exploited at maximum in a worst case scenario.

#### Fee Collection

This proposal entails collecting fees in ETH, rather than snxUSD, for the following reasons:

- Users will always receive the same amount of snxUSD that they have transferred.
- This is the most efficient method to collect fees, as no additional token transfers are involved in the transaction.
- Users are already paying ETH for gas fees to execute the transaction.
- Reduces code complexity since the `CcipRouter` handles the fee calculation for the execution of the specified message.

> - Users are automatically refunded any left over ETH that was unused in the message execution.

Reviewer: Suggest to clarify that this refund is specifically for the difference between the value that the user passes to transferCrossChain() and what CCIP Router returns as getFee().

> In this scenario, there is also potential to set up a rev-share model between CCIP and Synthetix (as token issuer) for third-party CCIP snxUSD token transfers, creating a shared incentive for increased snxUSD token transfer volume over CCIP.

Reviewer: ```suggestion

```

Future SIPs could involve additional fee payment methods (such as snxUSD), alternative fee structures, and circuit-breaking functionality for added security.

### Implementation

<!--The technical specification should outline the public API of the changes proposed. That is, changes to any of the interfaces Synthetix currently exposes or the creations of new ones.-->

Implementation entails adding two new functions to the snxUSD token module:

**`transferCrossChain(uint64 destChainId, address to, uint256 amount) external payable returns (uint256 gasTokenUsed)`**

- Confirm that the sender has a sufficient balance of snxUSD to transfer.

> - Encode the `destChainId`, `to` and `amount` with selector `mint` and call `ccipSend`.

Reviewer: To be reviewed: not fully clear what selector `mint` is used for. Is this relevant for CCIP Token Transfers?

Author: This has been updated in https://github.com/Synthetixio/SIPs/pull/1581/commits/58ab9952317c8daea9f41035bbdeac4fac435114 to reflect what is needed for CCIP Token Transfers (i.e. using token address and amount for `EVMTokenAmount[]`)

> - Burn the amount of snxUSD transferred.

Reviewer: ```suggestion
```

Reviewer: Burning the amount of snxUSD transferred is handled by CCIP Token Pool.

> - Refunds any left over gas token (i.e. ETH).

Reviewer: Suggest to clarify that this refund is specifically for the difference between the value that the user passes to transferCrossChain() and what CCIP Router returns as getFee(). See also comment above under 'Fee Collection'

**`ccipReceive(Any2EVMMessage calldata message)`**

- Verify that the `srcChainId` and `sender` data in the message correspond to a legitimate deployment of the snxUSD token.

> - Mint the amount of snxUSD tokens to the destination address, as specified in the message's `data`.

Reviewer: ```suggestion
```

Reviewer: Minting the amount of snxUSD to the `receiver` contract is handled by CCIP Token Pool. The `receiver` then transfers to the final recipient per the destination address, as specified in the message's `data`.

We propose making the transfer function available on the snxUSD proxy, rather than the core system proxy. This is an extension of the ERC-20 interface (which already essentially includes a "same-chain" `transfer` function).

#### Using CCIP for token transfers

In addition to the new functions outlined in the technical specification below, this SIP also proposes that Chainlink's CCIP token pools be granted permission to mint and burn snxUSD in order to facilitate the cross-chain token transfers. This method allows for multiple benefits as described below:

**Composability & Commercial Opportunity**

Token transfers using the CCIP token pools provide Synthetix the opportunity in the future, if desired, to open up Chainlink's private snxUSD Token Pool which would allow third parties to transfer snxUSD (even in combination with other tokens and/or data). In this scenario, there is also potential to set up a rev-share model between CCIP and Synthetix (as snxUSD token issuer) for third-party CCIP snxUSD token transfers, creating a shared incentive for increased snxUSD token transfer volume over CCIP.

**Enhanced security via the Active Risk Management (ARM) Network**

> With token transfers, the ARM Network has access to additional transaction information (i.e. token, amount) that enable it to perform additional risk management policies which cannot be implemented for arbitrary messaging. The ARM rule set will evolve over time into more advanced logic to identify and mitigate risks.

Reviewer: ```suggestion
With token transfers, the ARM Network has access to additional transaction information (i.e. token, amount) that enable it to perform additional risk management policies which cannot be implemented for arbitrary messaging. The ARM rule set will evolve over time into more advanced logic to identify and mitigate additional risks.
```

- This is the most efficient method to collect fees, as no additional token transfers are involved in the transaction.
- Users are already paying ETH for gas fees to execute the transaction.
- Reduces code complexity since the `CcipRouter` handles the fee calculation for the execution of the specified message.
- Users are automatically refunded the difference between the `msg.value` that the user passes to `transferCrossChain()` and what the `CcipRouter` returns as `getFee()`.

Future SIPs could involve additional fee payment methods (such as snxUSD), alternative fee structures, and circuit-breaking functionality for added security.

### Implementation

<!--The technical specification should outline the public API of the changes proposed. That is, changes to any of the interfaces Synthetix currently exposes or the creations of new ones.-->

Implementation entails adding two new functions to the snxUSD token module:

**`transferCrossChain(uint64 destChainId, address to, uint256 amount) external payable returns (uint256 gasTokenUsed)`**

- Confirm that the sender has a sufficient balance of snxUSD to transfer.

> - Encode the token addresses and amounts, and call `ccipSend`.

Reviewer: ```suggestion
Approve the amount of snxUSD to be transferred to the CCIP Router.
```

> - Approve the amount of snxUSD transferred to the CCIP Router.

Reviewer: ```suggestion
- Encode the snxUSD token address and amount, and call `ccipSend`.
```

<!--Test cases for an implementation are mandatory for SIPs but can be included with the implementation.-->

Relevant tests will be developed during implementation, including but not limited to:

#### Unit Tests

Two new unit tests are added to test the new contracts. These tests will mock the required external interfaces (i.e. `CcipRouter` and `CcipTokenPool`).

For `USDTokenModule`:

- It reverts transactions coming from an unauthorized address (i.e. non-CCIP address).
- It reverts if the sender does not have enough snxUSD balance or allowance.
- It gets deployed and registered correctly.
- It applies the expected USD parameters.

For `CcipReceiverModule`:

- It reverts if caller is not the `CcipRouter`.
- It reverts if chain is not supported.
- It reverts if message sender on other chain is not self.
- It forwards message to specified caller.

#### Integration Tests

One integration test to cover the initiation of a cross-chain transfer.

> - Verify it burns the correct amount of snxUSD on the source chain.

Reviewer: This can be replaced by verifying the allowance that the Router has been given via approve to transferFrom the amount of snxUSD transferred.
